ledger/eval/eval_test.go

```diff
@@ -1445,3 +1401,179 @@ func TestExpiredAccountGeneration(t *testing.T) {
+		{
+			actual:   basics.MicroAlgos{Raw: 99},
+			expected: basics.MicroAlgos{Raw: 100},
+			round:    basics.Round(config.Consensus[proto].StateProofInterval),
+			err:      "StateProofOnlineTotalWeight wrong: {99} != {100}",
+		},
+		// Exception round non-mainnet, mismatched state (actual>expected) => no error
+		{
+			actual:   basics.MicroAlgos{Raw: 100},
+			expected: basics.MicroAlgos{Raw: 99},
+			round:    basics.Round(27713280),
+			err:      "StateProofOnlineTotalWeight wrong: {100} != {99}",
+		},
+		// Exception round non-mainnet, mismatched state (actual<expected) => no error
+		{
+			actual:   basics.MicroAlgos{Raw: 99},
+			expected: basics.MicroAlgos{Raw: 100},
+			round:    basics.Round(27713280),
+			err:      "StateProofOnlineTotalWeight wrong: {99} != {100}",
+		},
+		// Exception round non-mainnet, mismatched stake with max possible value for muldiv => no error
+		{
+			actual:   basics.MicroAlgos{Raw: math.MaxUint64 - 100},
+			expected: basics.MicroAlgos{Raw: math.MaxUint64},
+			round:    basics.Round(27713280),
+			err:      "StateProofOnlineTotalWeight wrong: {18446744073709551515} != {18446744073709551615}",
+		},
+	}
+
+	for i, test := range tests {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			params := config.Consensus[proto]
+
+			genesisInitState, _, _ := ledgertesting.GenesisWithProto(2, proto)
+
+			l := newTestLedger(t, bookkeeping.GenesisBalances{
+				Balances:    genesisInitState.Accounts,
+				FeeSink:     testSinkAddr,
+				RewardsPool: testPoolAddr,
+				Timestamp:   0,
+			})
+			// If genesis hash is non-nil in test case, override the test hash here
+			if !test.genesisHash.IsZero() {
+				l.genesisHash = test.genesisHash
+			}
+			l.voters = map[basics.Round]*ledgercore.VotersForRound{
+				basics.Round(uint64(test.round) - params.StateProofVotersLookback): {
+					Tree:        &merklearray.Tree{},
+					TotalWeight: test.expected,
+				},
+			}
+
+			genesisBlockHeader, err := l.BlockHdr(0)
+			require.NoError(t, err)
+			newBlock := bookkeeping.MakeBlock(genesisBlockHeader)
+			newBlock.StateProofTracking = map[protocol.StateProofType]bookkeeping.StateProofTrackingData{
+				protocol.StateProofBasic: {
+					StateProofOnlineTotalWeight: test.actual,
+				},
+			}
+			newBlock.BlockHeader.Round = test.round
+
+			pcow := mockLedger{balanceMap: map[basics.Address]basics.AccountData{}}
+			mods := ledgercore.StateDelta{Hdr: &newBlock.BlockHeader}
+			cow := &roundCowState{
+				lookupParent: &pcow,
+				mods:         mods,
+			}
+			eval := BlockEvaluator{
+				state:       cow,
+				validate:    true,
+				prevHeader:  genesisBlockHeader,
+				proto:       params,
+				genesisHash: l.genesisHash,
+				block:       newBlock,
+				l:           l,
+			}
+
+			err = eval.endOfBlock()
+			if len(test.err) == 0 {
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), test.err)
```

Comment:
nit/fyi: `require.ErrorContains` could be used to consoldate these two checks

Reply:
Agreed!